src/app/map/shared/layer.service.ts

```diff
@@ -12,13 +15,63 @@ export class LayerService {
     osm: OSMLayer,
     vector: VectorLayer,
     xyz: XYZLayer,
+    wmts: WMTSLayer
   };
 
-  constructor() { }
+  getCapabilitiesStore: any[] = [];
+
+  constructor(private http: Http) { }
+
+  createLayer(options: LayerOptions): Observable<Layer> {
 
-  createLayer(options: LayerOptions): Layer {
     const layerCls = LayerService.layerClasses[options.type];
 
-    return new layerCls(options);
+    if (options.optionsFromCapabilities) {
+       return this.getCapabilities(options).map(
+        getCapabilities => {
+          return new layerCls(options, getCapabilities);
+        });
+    } else {
+      return new Observable(layer => layer.next(new layerCls(options)));
+    }
+  }
+
+  getCapabilities(options): Observable<any> {
+    let url = options.source.url + '?REQUEST=GetCapabilities&';
+    url += 'SERVICE=' + options.type + '&';
+    url += 'VERSION=' + (options.version ? options.version : '1.0.0') + '&';
+
+    const getCapabilities = this.findGetCapabilities(url);
+    if (getCapabilities) {
+     return getCapabilities.getCapabilities;
+    } else {
+      return this.http.get(url)
+              .map(response => {
+
+                  let parser;
+                  switch (options.type) {
+                    case 'wmts':
+                      parser = new ol.format.WMTSCapabilities();
+                      break;
+                    case 'wms':
+                      parser = new ol.format.WMSCapabilities();
+                      break;
+                  }
+
+                  getCapabilities = parser.read(response.text());
+                  this.getCapabilitiesStore.push({url: url, getCapbilities: getCapabilities});
+                  return getCapabilities;
+              });
+    }
+  }
+
+  findGetCapabilities(url: string): any {
+
+    return this.getCapabilitiesStore.find(function(value, idx, array){
```

Comment:
You can simplify this by writting `this.getCapabilitiesStore.find(value => value.url === url);`